### PR TITLE
ast: Initial implementation of abstract syntax tree.

### PR DESCRIPTION
**Pull request by coocood**
In order to do better optimization, we should refactor parser first.

The code is not used, but ready to be reviewed and merged.

### CODE REVIEW

File: ast/ast.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+// ExprNode is a node that can be evaluated.
+// Name of implementations should have 'Expr' suffix.
+type ExprNode interface {
+	// Node is embeded in ExprNode.
+	Node
+	// IsStatic means it can be evaluated independently.
+	IsStatic() bool
+	// SetType sets type to the expression.
```
Comment (siddontang) [inline ast/ast.go]: return type

File: tidb.go [hunk 2]
```diff
@@ -92,8 +94,20 @@ func Compile(src string) ([]stmt.Statement, error) {
 		log.Warnf("compiling %s, error: %v", src, l.Errors()[0])
 		return nil, errors.Trace(l.Errors()[0])
 	}
-
-	return l.Stmts(), nil
+	rawStmt := l.Stmts()
+	stmts := make([]stmt.Statement, len(rawStmt))
+	for i, v := range rawStmt {
+		if node, ok := v.(ast.Node); ok {
+			stm, err := optimizer.Compile(node)
```
Comment (qiuyesuifeng) [inline tidb.go]: stm -> stmt?

Comment (coocood) [inline tidb.go]: I think it's ok for local variable.

File: ast/ast.go (hunk 1 above)
Comment (ngaut) [inline ast/ast.go]: s/to/of ?

File: ast/base.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+// limitations under the License.
+
+package ast
+
+import "github.com/pingcap/tidb/util/types"
+
+// node is the struct implements node interface except for Accept method.
+// Node implementations should embed it in.
+type node struct {
+	txt string
```
Comment (ngaut) [inline ast/base.go]: s/txt/text

**Comment by ngaut** [comment]
LGTM

**Pull request merged**